CTClassifier fix

gavinmischler opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/mvlearn/mvlearn/blob/main/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #317 

#### What does this implement/fix? Explain your changes.
Changes the attribute names of estimator1 and estimator2 in CTClassifier and CTRegressor to not have trailing underscores, which allows them to be properly picked up by sklearn's get_params() method. Also editted other init args which were set as attributes with trailing underscores.

#### Any other comments?

gavinmischler committed:
fix for ctclassifier and ctregressor attributes

gavinmischler committed:
increment version

gavinmischler committed:
remove old python version from tests since not supported by github actions anymore

gavinmischler committed:
fix tests

agramfort commented on mvlearn/semi_supervised/ctclassifier.py:
```diff
@@ -178,20 +178,20 @@ def __init__(self,
-            self.estimator2_ = GaussianNB()
+        if self.estimator1 is None:
+            self.estimator1 = GaussianNB()
+        if self.estimator2 is None:
+            self.estimator2 = GaussianNB()
 
         # If only 1 of p or n is not None, set them equal
         if (p is not None and n is None):
             n = p
-            self.p_, self.n_ = p, n
+            self.p, self.n = p, n
         elif (p is None and n is not None):
             p = n
-            self.p_, self.n_ = p, n
+            self.p, self.n = p, n
         else:
-            self.p_, self.n_ = p, n
+            self.p, self.n = p, n
```
this should be done in the fit. There should be no logic in the init just do:

self.a = a
self.b = b

etc.

agramfort commented on mvlearn/semi_supervised/ctclassifier.py:
```diff
@@ -305,30 +305,30 @@ def fit(self, Xs, y):
-                self.estimator1_.fit(X1[L], y[L])
-                self.estimator2_.fit(X2[L], y[L])
+                self.estimator1.fit(X1[L], y[L])
+                self.estimator2.fit(X2[L], y[L])
```
the fit should not change the init parameters but just add / modify attributes that end with _. So you should create 2 new attributes here called estimator1_ and estimator2_ (on top of estimator1 and estimator2)

agramfort commented on mvlearn/semi_supervised/ctclassifier.py:
```diff
@@ -271,14 +271,14 @@ def fit(self, Xs, y):
         if self.n_classes > 1:
 
             # if both p & n are none, set as ratio of one class to the other
-            if (self.p_ is None and self.n_ is None):
```
also I would avoid single letter attributes. Use n_samples_ and n_features_ ?

gavinmischler replied:
Changed them to ``positive_samples`` and ``negative_samples``

gavinmischler committed:
fix pycodestyle

gavinmischler committed:
add underscores to estimator1-2 and lengthen attribute names

gavinmischler committed:
update docker convenience image since old images depracated

gavinmischler committed:
change docker image

gavinmischler commented:
@agramfort I believe I've addressed your comments. Regarding the tests failing, I opened another PR because the tests are from other dependencies having things deprecated, unrelated to CTClassifier. If we merge that one first it will fix the tests here.

Also, regarding the CircleCI docs build failing, it's failing because CircleCI deprecated the python docker image we were using, but when I tried updating to a new image (``cimg/python:3.18.15``) it still doesn't work. Do you have experience with this?

agramfort commented on mvlearn/semi_supervised/ctregression.py:
```diff
@@ -278,8 +276,8 @@ def fit(self, Xs, y):
-        self.estimator1_.fit(X1[L1], y1[L1])
-        self.estimator2_.fit(X2[L2], y2[L2])
+        self.estimator1.fit(X1[L1], y1[L1])
+        self.estimator2.fit(X2[L2], y2[L2])
```
this was correct. See self.estimator1 as a kind of factory that you can copy/clone to get an instance you want to use at fit time. The fit method should only modify parameters that end with _. The self.estimator1 should stay untouched. It means you need both self.estimator1 and self.estimator1_ (the clean one from init you could reuse for a new fit on a new dataset) and the one used at fit one the last data.

gavinmischler replied:
got it, I had made the change in ctclassifier but not ctregression. Now I make a copy of the original estimator1 and estimator2 each time ``fit`` is called

gavinmischler committed:
copy estimators in fit

agramfort reviewed (approved):
LGTM

no idea how hard it is at this point to make the CIs green

gavinmischler committed:
update circleci config to version 2.1

gavinmischler committed:
change to python orb

gavinmischler committed:
change xvfb

gavinmischler committed:
print which xvfb

gavinmischler committed:
print which xvfb

gavinmischler committed:
remove xvfb

gavinmischler committed:
remove qt5 install

gavinmischler committed:
add back sudo apt-get update

gavinmischler committed:
fix examples

gavinmischler committed:
update dcca and groupica test

gavinmischler committed:
fix pycodestyle for ctclassifier

gavinmischler committed:
fix broken tests

gavinmischler committed:
fix examples

gavinmischler commented:
I got the docs to build at least and the tests pass, but there seems to be an issue with accessing the built docs artifact. I wonder if this is related to [this PR](https://github.com/mne-tools/mne-python/pull/11622) in MNE which added a CircleCI API token for their CircleCI Artifacts workflow. I don't have the permissions for mvlearn or circleci to create such a token and add it as a secret though, so I don't have a way of testing this.

Also, someone with permissions should remove the PR checklist requirement of the 3.6 tests because this PR removed those in favor of the 3.7 and 3.8 tests but it's still limiting whether the PR can be merged
